### ISSUE TITLE
Instance resolution loops infinitely even when an instance is available

### ISSUE THREAD

**Issue by beren-evans-oguz** [agda/agda#4924]
On `Agda version 2.6.1.1` typechecking this program loops infinitely:

```agda
{-# OPTIONS --safe #-}

module Bad where

record A (T : Set) : Set where
  field
    someT : T
open A {{...}}

record B (T : Set) (a : A T) : Set where
  field
    f : T → T

  instance
    aInstance : A T
    aInstance = a
open B

bad : ∀ {T : Set} {a : A T} → B T a → T
bad b = let open B b in f someT
```

But it was pointed out in IRC channel by some other user (`tomjack`) that this works, typechecking halts:

```agda
{-# OPTIONS --safe #-}

module Good where

record A (T : Set) : Set where
  field
    someT : T
open A {{...}}

record B (T : Set) (a : A T) : Set where
  field
    f : T → T

  instance
    aInstance : A T
    aInstance = a
open B

good : ∀ {T : Set} {a : A T} → B T a → T
good b = let open B b renaming (f to fb) in fb someT
```

Same issue exists if one uses `where` instead of `let` clause.

**Comment by tomjack** [comment]
Maybe it is worth noting also that the problem appears even with `--instance-search-depth=1`. So I don't know what is looping infinitely...

With some verbosity you see this repeated at the end:
```
no displayForm for Bad._.aInstance
reifying def Bad._.aInstance
no displayForm for Bad._.aInstance
reifying def Bad._.aInstance
...
```

**Comment by beren-evans-oguz** [comment]
Removing the `open B` in the first program makes agda halt. This typechecks:

```agda
{-# OPTIONS --safe #-}

module Good2 where

record A (T : Set) : Set where
  field
    someT : T
open A {{...}}

record B (T : Set) (a : A T) : Set where
  field
    f : T → T

  instance
    aInstance : A T
    aInstance = a

good2 : ∀ {T : Set} {a : A T} → B T a → T
good2 b = let open B b in f someT
```

notice that there is no `open B` after `record B`. This makes sense. Without instance search, I can write a program that is rejected by typechecker:

```agda
-- This program does not typecheck
{-# OPTIONS --safe #-}

module Good3 where

record A (T : Set) : Set where
  field
    someT : T
open A {{...}}

record B (T : Set) (a : A T) : Set where
  field
    f : T

  instance
    aInstance : A T
    aInstance = a
open B

good3 : ∀ {T : Set} {a : A T} → B T a → T
good3 b = let open B b in f
```

Typechecking fails with the following error:

```
Cannot resolve overloaded projection f because it is not applied to
a visible argument
when checking that the expression f has type T
```

It seems to me that agda wants to express that "`open B b` shadows `f` that was previously exposed by `open B`". But it being applied to an instance argument causes agda to loop. As pointed out earlier by @tomjack it doesn't seem like instance search per se is looping, but the need for instance resolution seems to trigger the bug here...

**Comment by uncle-betty** [comment]
I managed to collect a little more data on this one with current `master`. This seems to be about the constraint solver.

What's recursing forever is `solveSomeAwakeConstraintsTCM.solve`. I wrapped the function into `verboseBracket` and observed the following evaluation sequence once we're looping forever. `verboseBracket` nesting depths / recursion depths are represented by indentation. `A`, `B`, ... are placeholders for constraint arguments. They are detailed below.

```
... previous evaluations of solve, all of which complete... 
solve [A]
    solve [B]
        solve [C]
            solve []
        solve [D]
            solve []
        solve []
        solve []
    solve [A']
        solve [B']
... looping forever...
```

The sequence starts repeating with `A'`, `B'`, ... where `A'`, `B'`, ... are almost completely identical to `A`, `B`, ... They differ only in their `MetaId`s, which are increased by 4. So, we are caught in an endless loop where only the `MetaId`s in the constraints keep changing.

The following are the constraints. I generated them from the very first of the above examples.

The following is `A`. It's the biggest one. To shorten things a little, I summarized `QName`s and removed `Range`s. Note the `MetaId 11` in there. It changes to `MetaId 15` in `A'`, which is the only difference between `A` and `A'`.

```
[
  PConstr {
    constraintProblems = fromList [
    ], constraintUnblocker = UnblockOnAny (fromList [
    ]), theConstraint = Closure {
       clValue = FindInstance (MetaId 11) (Just [
        Candidate {
          candidateKind = GlobalCandidate (QName {
             ### "Bad.B.aInstance" ###
          }), candidateTerm = Lam (ArgInfo {
            argInfoHiding = Hidden, argInfoModality = Modality {
              modRelevance = Relevant, modQuantity = Quantityω QωInferred, modCohesion = Continuous
            }, argInfoOrigin = Inserted, argInfoFreeVariables = UnknownFVs, argInfoAnnotation = Annotation {
              annLock = IsNotLock
            }
          }) (Abs "T" Lam (ArgInfo {
            argInfoHiding = Hidden, argInfoModality = Modality {
              modRelevance = Relevant, modQuantity = Quantityω QωInferred, modCohesion = Continuous
            }, argInfoOrigin = Inserted, argInfoFreeVariables = UnknownFVs, argInfoAnnotation = Annotation {
              annLock = IsNotLock
            }
          }) (Abs "a" Lam (ArgInfo {
            argInfoHiding = NotHidden, argInfoModality = Modality {
              modRelevance = Relevant, modQuantity = Quantityω QωInferred, modCohesion = Continuous
            }, argInfoOrigin = Inserted, argInfoFreeVariables = UnknownFVs, argInfoAnnotation = Annotation {
              annLock = IsNotLock
            }
          }) (Abs "r" Var 1 [
          ]))), candidateType = El {
            _getSort = Type (Max 1 [
            ]), unEl = Pi (Dom {
              domInfo = ArgInfo {
                argInfoHiding = Hidden, argInfoModality = Modality {
                  modRelevance = Relevant, modQuantity = Quantityω QωInferred, modCohesion = Continuous
                }, argInfoOrigin = UserWritten, argInfoFreeVariables = UnknownFVs, argInfoAnnotation = Annotation {
                  annLock = IsNotLock
                }
              }, domFinite = False, domName = Just (WithOrigin {
                woOrigin = Inserted, woThing = Ranged {
                  rangeOf = ...,
                  rangedThing = "T"
                }
              }), domTactic = Nothing, unDom = El {
                _getSort = Type (Max 1 [
                ]), unEl = Sort (Type (Max 0 [
                ]))
              }
            }) (Abs "T" El {
              _getSort = Type (Max 0 [
              ]), unEl = Pi (Dom {
                domInfo = ArgInfo {
                  argInfoHiding = Hidden, argInfoModality = Modality {
                    modRelevance = Relevant, modQuantity = Quantityω QωInferred, modCohesion = Continuous
                  }, argInfoOrigin = UserWritten, argInfoFreeVariables = UnknownFVs, argInfoAnnotation = Annotation {
                    annLock = IsNotLock
                  }
                }, domFinite = False, domName = Just (WithOrigin {
                  woOrigin = Inserted, woThing = Ranged {
                    rangeOf = ...,
                    rangedThing = "a"
                  }
                }), domTactic = Nothing, unDom = El {
                  _getSort = Type (Max 0 [
                  ]), unEl = Def (QName {
                    ### "Bad.A" ###
                  }) [
                    Apply (Arg {
                      argInfo = ArgInfo {
                        argInfoHiding = NotHidden, argInfoModality = Modality {
                          modRelevance = Relevant, modQuantity = Quantityω QωInferred, modCohesion = Continuous
                        }, argInfoOrigin = UserWritten, argInfoFreeVariables = UnknownFVs, argInfoAnnotation = Annotation {
                          annLock = IsNotLock
                        }
                      }, unArg = Var 0 [
                      ]
                    })
                  ]
                }
              }) (Abs "a" El {
                _getSort = Type (Max 0 [
                ]), unEl = Pi (Dom {
                  domInfo = ArgInfo {
                    argInfoHiding = NotHidden, argInfoModality = Modality {
                      modRelevance = Relevant, modQuantity = Quantityω QωInferred, modCohesion = Continuous
                    }, argInfoOrigin = UserWritten, argInfoFreeVariables = UnknownFVs, argInfoAnnotation = Annotation {
                      annLock = IsNotLock
                    }
                  }, domFinite = False, domName = Nothing, domTactic = Nothing, unDom = El {
                    _getSort = Type (Max 0 [
                    ]), unEl = Def (QName {
                      ### "Bad.B" ###
                    }) [
                      Apply (Arg {
                        argInfo = ArgInfo {
                          argInfoHiding = NotHidden, argInfoModality = Modality {
                            modRelevance = Relevant, modQuantity = Quantityω QωInferred, modCohesion = Continuous
                          }, argInfoOrigin = UserWritten, argInfoFreeVariables = UnknownFVs, argInfoAnnotation = Annotation {
                            annLock = IsNotLock
                          }
                        }, unArg = Var 1 [
                        ]
                      }),Apply (Arg {
                        argInfo = ArgInfo {
                          argInfoHiding = NotHidden, argInfoModality = Modality {
                            modRelevance = Relevant, modQuantity = Quantityω QωInferred, modCohesion = Continuous
                          }, argInfoOrigin = UserWritten, argInfoFreeVariables = UnknownFVs, argInfoAnnotation = Annotation {
                            annLock = IsNotLock
                          }
                        }, unArg = Var 0 [
                        ]
                      })
                    ]
                  }
                }) (Abs "r" El {
                  _getSort = Type (Max 0 [
                  ]), unEl = Def (QName {
                    ### "Bad.A" ###
                  }) [
                    Apply (Arg {
                      argInfo = ArgInfo {
                        argInfoHiding = NotHidden, argInfoModality = Modality {
                          modRelevance = Relevant, modQuantity = Quantityω QωInferred, modCohesion = Continuous
                        }, argInfoOrigin = UserWritten, argInfoFreeVariables = UnknownFVs, argInfoAnnotation = Annotation {
                          annLock = IsNotLock
                        }
                      }, unArg = Var 2 [
                      ]
                    })
                  ]
                })
              })
            })
          }, candidateOverlappable = False
        },Candidate {
          candidateKind = GlobalCandidate (QName {
            ### "Bad._.aInstance" ###
          }), candidateTerm = Var 1 [
          ], candidateType = El {
            _getSort = Type (Max 0 [
            ]), unEl = Def (QName {
              ### "Bad.A" ###
            }) [
              Apply (Arg {
                argInfo = ArgInfo {
                  argInfoHiding = NotHidden, argInfoModality = Modality {
                    modRelevance = Relevant, modQuantity = Quantityω QωInferred, modCohesion = Continuous
                  }, argInfoOrigin = UserWritten, argInfoFreeVariables = UnknownFVs, argInfoAnnotation = Annotation {
                    annLock = IsNotLock
                  }
                }, unArg = Var 2 [
                ]
              })
            ]
          }, candidateOverlappable = False
        }
      ])
    }
  }
]
```

`B` looks like this.

```
[
  PConstr {
    constraintProblems = fromList [
    ], constraintUnblocker = UnblockOnMeta (MetaId 10), theConstraint = Closure {
       clValue = UnBlock (MetaId 12)
    }
  }
]
```

`C` looks like this. Note that this is the fist time we see `MetaId 15`, i.e., the `MetaId` that'll show up in `A'`.

```
[
  PConstr {
    constraintProblems = fromList [
    ], constraintUnblocker = UnblockOnAll (fromList [
    ]), theConstraint = Closure {
       clValue = FindInstance (MetaId 15) Nothing
    }
  }
]
```

`D` is longer again.

```
[
  PConstr {
    constraintProblems = fromList [
    ], constraintUnblocker = UnblockOnMeta (MetaId 12), theConstraint = Closure {
       clValue = ValueCmp CmpEq (AsTermsOf (El {
        _getSort = Type (Max 0 [
        ]), unEl = Var 2 [
        ]
      })) (MetaV (MetaId 13) [
        Apply (Arg {
          argInfo = ArgInfo {
            argInfoHiding = Hidden, argInfoModality = Modality {
              modRelevance = Relevant, modQuantity = Quantityω QωInferred, modCohesion = Continuous
            }, argInfoOrigin = UserWritten, argInfoFreeVariables = UnknownFVs, argInfoAnnotation = Annotation {
              annLock = IsNotLock
            }
          }, unArg = Var 2 [
          ]
        }),Apply (Arg {
          argInfo = ArgInfo {
            argInfoHiding = Hidden, argInfoModality = Modality {
              modRelevance = Relevant, modQuantity = Quantityω QωInferred, modCohesion = Continuous
            }, argInfoOrigin = UserWritten, argInfoFreeVariables = UnknownFVs, argInfoAnnotation = Annotation {
              annLock = IsNotLock
            }
          }, unArg = Var 1 [
          ]
        }),Apply (Arg {
          argInfo = ArgInfo {
            argInfoHiding = NotHidden, argInfoModality = Modality {
              modRelevance = Relevant, modQuantity = Quantityω QωInferred, modCohesion = Continuous
            }, argInfoOrigin = UserWritten, argInfoFreeVariables = UnknownFVs, argInfoAnnotation = Annotation {
              annLock = IsNotLock
            }
          }, unArg = Var 0 [
          ]
        })
      ]) (MetaV (MetaId 12) [
        Apply (Arg {
          argInfo = ArgInfo {
            argInfoHiding = Hidden, argInfoModality = Modality {
              modRelevance = Relevant, modQuantity = Quantityω QωInferred, modCohesion = Continuous
            }, argInfoOrigin = UserWritten, argInfoFreeVariables = UnknownFVs, argInfoAnnotation = Annotation {
              annLock = IsNotLock
            }
          }, unArg = Var 2 [
          ]
        }),Apply (Arg {
          argInfo = ArgInfo {
            argInfoHiding = Hidden, argInfoModality = Modality {
              modRelevance = Relevant, modQuantity = Quantityω QωInferred, modCohesion = Continuous
            }, argInfoOrigin = UserWritten, argInfoFreeVariables = UnknownFVs, argInfoAnnotation = Annotation {
              annLock = IsNotLock
            }
          }, unArg = Var 1 [
          ]
        }),Apply (Arg {
          argInfo = ArgInfo {
            argInfoHiding = NotHidden, argInfoModality = Modality {
              modRelevance = Relevant, modQuantity = Quantityω QωInferred, modCohesion = Continuous
            }, argInfoOrigin = UserWritten, argInfoFreeVariables = UnknownFVs, argInfoAnnotation = Annotation {
              annLock = IsNotLock
            }
          }, unArg = Var 0 [
          ]
        })
      ])
    }
  }
]
```

Is there a place or paper where I can read up on Agda's constraint solver?

**Comment by uncle-betty** [comment]
Oh, I just discovered `debugConstraints`, which concisely pretty-prints all active constraints, including any sleeping constraints. That's more manageable and more informative, I guess.

This is pretty-printed `A` (and any sleeping constraints when entering `solve` for `A`).

```
Current constraints
  awake  Resolve instance argument
           _r_11
             : {T = T₁ : Set} {a = a₁ : A T₁} (b₁ : B T₁ a₁) →
               A (_T_10 (b = b₁))
         Candidates
           B.aInstance : {T = T₁ : Set} {a = a₁ : A T₁} (r : B T₁ a₁) → A T₁
           Bad.aInstance : A T
           (stuck)
  asleep _12 := f someT :? T (blocked on _T_10)
```

Same for `B`.

```
Current constraints 
  awake  _12 := f someT :? T (blocked on _T_10) 
  asleep 
```

Same for `C`.

```
Current constraints
  awake  Resolve instance argument
           _r_15
             : {T = T₁ : Set} {a = a₁ : A T₁} (b₁ : B T₁ a₁) →
               A (_T_14 (b = b₁))
         No candidates yet
           ()
  asleep _16 := f someT :? T (blocked on _T_14)
```

Same for `D`.

```
Current constraints
  awake  _13 = _17 : T (blocked on _12)
  asleep Resolve instance argument
           _r_15
             : {T = T₁ : Set} {a = a₁ : A T₁} (b₁ : B T₁ a₁) →
               A (_T_14 (b = b₁))
         Candidates
           B.aInstance : {T = T₁ : Set} {a = a₁ : A T₁} (r : B T₁ a₁) → A T₁
           Bad.aInstance : A T
           (stuck)
         _16 := f someT :? T (blocked on _T_14)
```

The following debug output covers the involved meta variables `_10` through `_17` being created. It seems like this could be useful data as well.

```
new meta (ConcreteDef): [{_}, {_}, b] |-
  _10 : {T = T₁ : Set} {a = a₁ : A T₁} (b₁ : B T₁ a₁) → Set
setting name of meta _10 to T

new instance meta: [{_}, {_}, b] |-
  _11 : {T = T₁ : Set} {a = a₁ : A T₁} (b₁ : B T₁ a₁) →

new meta _12 : {T : Set} {a : A T} (b : B T a) → T
for postponed typechecking problem f someT :? @2

new meta (ConcreteDef): [{_}, {_}, b] |-
  _13 : {T = T₁ : Set} {a = a₁ : A T₁} (b₁ : B T₁ a₁) → T₁

new meta (ConcreteDef): [{_}, {_}, b] |-
  _14 : {T = T₁ : Set} {a = a₁ : A T₁} (b₁ : B T₁ a₁) → Set
setting name of meta _14 to T

new meta (ConcreteDef): [{_}, {_}, b] |-
  _15 : {T = T₁ : Set} {a = a₁ : A T₁} (b₁ : B T₁ a₁) →
        A (_T_14 (b = b₁))

new meta (ConcreteDef): [{_}, {_}, b] |-
  _14 : {T = T₁ : Set} {a = a₁ : A T₁} (b₁ : B T₁ a₁) → Set
setting name of meta _14 to T

new instance meta: [{_}, {_}, b] |-
  _15 : {T = T₁ : Set} {a = a₁ : A T₁} (b₁ : B T₁ a₁) →
        A (_T_14 (b = b₁))

new meta _16 : {T : Set} {a : A T} (b : B T a) → T
for postponed typechecking problem f someT :? @2

new meta (ConcreteDef): [{_}, {_}, b] |-
  _17 : {T = T₁ : Set} {a = a₁ : A T₁} (b₁ : B T₁ a₁) → T₁
```

**Comment by uncle-betty** [comment]
I think that I now understand what's going on in `Bad`.

`open B` yields the following types.

```
f         : {T : Set} {a : A T} -> (B T a) -> T -> T
aInstance : {T : Set} {a : A T} -> (B T a) -> A T
```

`open B b` yields the following types, where `T` is `b`'s `T`.

```
f         : T -> T
aInstance : A T
```

We run into trouble when type checking `f someT`, i.e., an application of an ambiguous projection, `f`. The type of `someT` looks as follows.

```
someT : {T : Set} {{r : A T}} -> T
```

To figure out which `f` to chose, Agda considers the type of the first argument that `f` is applied to - the *principal argument*. Here it's the type of `someT`. This type is required to be a record type. Once `someT`'s implicits / instances are resolved, however, its type is just `T`. So, we'd expect Agda to error out saying that `someT`'s type is not a record type. Instead, Agda loops forever. Why?

Agda needs to figure out the type of `someT`. As a first step, she creates meta variables for the implicit / instance aguments of `someT`, i.e., for `{T : Set}` and `{{r : A T}}`. For `Bad`, the meta variables are `_10` and `_11`. They look like this.

```
_10 : {T = T₁ : Set} {a = a₁ : A T₁} (b₁ : B T₁ a₁) -> Set
_11 : {T = T₁ : Set} {a = a₁ : A T₁} (b₁ : B T₁ a₁) -> A (_T_10 (b = b₁))
```

Note that Agda included the types of the arguments of `bad`. I don't yet understand this. I guess it's just the context in which Agda tries to figure out the meta variables. In any case, telling from the type checking code, these meta variables do indeed come from `someT`'s hidden (implicit / instance) arguments.

Let's ignore `bad`'s arguments for now, as they are immaterial for the cause of this issue. This leaves us with `Set` and `A (_T_10)` as the meta variable types. This is appropriate, because:

* `_10` is `T` and `T` is a `Set`.
* `_11` is `r` and `r` is an `A T`. The fact that we're talking about the `T` given by `_10` is expressed by `_T_10`.

Once we have `_10` and `_11`, we have the type of `someT`, which is just `T`, which is to be found in `_10`. As we don't yet know `_10`, the type check of `f someT` needs to be *postponed* until after we have `_10`.

This delay is expressed by the next generated meta variable, `_12`. It captures the current state of the type check, so that we can pick up where we left off once we have `_10`. Postponing also involves a second meta variable, `_13`. I don't yet understand its purpose, but we don't need it to get to the bottom of this issue.

Agda now figures out `_10` and `_11` - which then makes her return to the postponed type check problem captured by meta variable `_12`. This is when the issue happens. Instead of finishing the type check based on `_10` and `_11`, Agda starts over and creates new meta variables `_14` through `_17` as replacements for `_10` through `_13`.

So, the whole thing starts over: Agda has to figure out `_14` and `_15`, then resumes the postponed type check problem `_16` - just to start over once more with new meta variables `_18` through `_21`. And so on. That's the loop. That's also why we see meta variable IDs increase by 4 between "loop iterations."

The above branch on my fork contains a proposed fix. It makes Agda remember meta variables `_10` and `_11` when resuming the type check, so that it can be finished with the now known values for `_10` and `_11`.

The proposed fix makes Agda indeed complain that `someT`'s type is not a record type when type checking `Bad`.

It was a fun exercise to explore and study the type checker a little while trying to figure this out, but I'm in way over my head. While the fix seems to work, I'll have to do more testing and then it would be great, if somebody more familiar with Agda could take a look. Maybe I'm only fixing a symptom of a different underlying issue. I can't tell for sure.

**Comment by uncle-betty** [comment]
Based on the above insights, this is a slightly simpler example to reproduce the issue:

```agda
module Repro where

record R (T : Set) : Set where
  field
    f : T → T

module _ {A : Set} (r : R A) (pri : {T : Set} → {{r : R T}} → T) where
  open R
  open R r

  instance
    _ = r

  foo : A
  foo = f pri
```

These factors seem to have to come together:

* An ambiguous projection, `f`. (The ambiguity resulting from opening `R` in two different ways.)
* A principal argument, `pri`
  * whose type, `T`, depends on an implicit argument, so that the need for resolving this argument postpones the type check of the application.
  * that takes an instance argument, `r`, which triggers instance resolution, which fills in the meta variables for the hidden arguments, which triggers pick-up of the postponed type check.
* An instance that makes instance resolution succeed.

In not so good news, it turns out that the fix I'm contemplating breaks `f ?`, i.e., using holes as principal arguments. It looks like I still don't fully understand how ambiguous projections are resolved.

**Comment by uncle-betty** [comment]
Turns out that the problem with `f ?` was a bug in my proposed fix. I'll create a pull request.

**Comment by andreasabel** [comment]
Great work, thanks, @uncle-betty !
